### PR TITLE
Improve constructors and variants

### PR DESCRIPTION
**Pull request by jlbabilino**
Redundant constructors and variant wrappers have been deleted.

C++'s implicitly defined constructors could accomplish exactly what the user-defined ones did, so they were removed.

The variant wrappers originally were intended to abstract an implementation detail and prevent the user from changing the variant's type after construction. I think these benefits don't warrant the amount of boilerplate code added. Eventually, there will be a `PathBuilder` class which completely abstracts the data structure implementation anyways (for example: `builder.ApplyStartPose(1_m, 2_m, 4_rad)`).

### CODE REVIEW

File: include/util/AppendVariant.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) TrajoptLib contributors
+
+#pragma once
+
+#include <variant>
+
+// NOLINTBEGIN
```
Comment (calcmogul) [inline include/util/AppendVariant.h:7]: Why are there NOLINT tags here?

File: include/util/AppendVariant.h [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+template <typename... Vs, typename... Aps>
+std::variant<Vs..., Aps...> _append_variant(std::variant<Vs...> myvar,
```
Comment (calcmogul) [inline include/util/AppendVariant.h:9]: You could name this variant_cat like https://en.cppreference.com/w/cpp/utility/tuple/tuple_cat.

File: include/util/AppendVariant.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+                                            Aps... newvals) {
+  return std::variant<Vs..., Aps...>();
```
Comment (calcmogul) [inline include/util/AppendVariant.h:11]: This returns a default-constructed variant instead of `myvar` or `newvals`.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: include/util/AppendVariant.h (hunk 3 above)
Comment (jlbabilino) [reply include/util/AppendVariant.h:11]: If you look at the way it's used, the function body doesn't matter. It's used in a `decltype()` block which iirc doesn't get value-evaluated. The point of the function isn't to take two variants and create a new one with the combined contents, it just gets the type that contains the types of the variant and other types listed.

File: include/util/AppendVariant.h (hunk 2 above)
Comment (jlbabilino) [reply include/util/AppendVariant.h:9]: I don't think this function is similar enough to that one. Unlike a tuple, a variant only holds one value, so it doesn't really make sense to make a new one that concatenates them.

File: include/util/AppendVariant.h (hunk 3 above)
Comment (calcmogul) [reply include/util/AppendVariant.h:11]: The behavior and intended usage should be documented.

File: include/util/AppendVariant.h (hunk 1 above)
Comment (jlbabilino) [reply include/util/AppendVariant.h:7]: `clang-tidy` will get angry about the fact that the parameters aren't used. The parameters are used but only for template type deduction.

File: include/util/AppendVariant.h (hunk 3 above)
Comment (jlbabilino) [reply include/util/AppendVariant.h:11]: Ok, on it

File: include/util/AppendVariant.h (hunk 1 above)
Comment (calcmogul) [reply include/util/AppendVariant.h:7]: Add the `[[maybe_unused]]` attribute to the parameters.

Comment (jlbabilino) [reply include/util/AppendVariant.h]: I've added docs and `[[maybe_unused]]`

**Review by calcmogul — APPROVED** [review]

**Pull request merged**